trie/zk_merkle_state_tree.go

```diff
@@ -0,0 +1,160 @@
+	return z.get(key)
+}
+
+func (z *ZkMerkleStateTree) GetAccount(address common.Address) (*types.StateAccount, error) {
+	if blob, err := z.get(address[:]); blob == nil || err != nil {
+		return nil, err
+	} else {
+		return types.UnmarshalStateAccount(blob)
+	}
+}
+
+func (z *ZkMerkleStateTree) get(key []byte) ([]byte, error) {
+	if hash, err := z.hash(key); err != nil {
+		return nil, err
+	} else {
+		return z.MerkleTree.Get(hash[:])
+	}
+}
+
+func (z *ZkMerkleStateTree) UpdateStorage(_ common.Address, key, value []byte) error {
+	return z.update(key, value)
+}
+
+func (z *ZkMerkleStateTree) UpdateAccount(address common.Address, account *types.StateAccount) error {
+	values, _ := account.MarshalFields()
+	var v []byte
+	for _, value := range values {
+		v = append(v, value.Bytes()...)
+	}
+	return z.update(address[:], v)
+}
+
+func (z *ZkMerkleStateTree) UpdateContractCode(_ common.Address, _ common.Hash, _ []byte) error {
+	return nil
+}
+
+// Update does not hash the key.
+func (z *ZkMerkleStateTree) Update(key, value []byte) error {
+	return z.MerkleTree.Update(key, value)
+}
+
+func (z *ZkMerkleStateTree) update(key, value []byte) error {
+	if hash, err := z.hash(key); err != nil {
+		return err
+	} else {
+		return z.MerkleTree.Update(hash[:], value)
+	}
+}
+
+func (z *ZkMerkleStateTree) DeleteStorage(_ common.Address, key []byte) error { return z.delete(key) }
+func (z *ZkMerkleStateTree) DeleteAccount(address common.Address) error       { return z.delete(address[:]) }
+func (z *ZkMerkleStateTree) delete(key []byte) error {
+	if hash, err := z.hash(key); err != nil {
+		return err
+	} else {
+		return z.MerkleTree.Delete(hash[:])
+	}
+}
+
+func (z *ZkMerkleStateTree) Hash() common.Hash {
+	hash, _, _ := z.Commit(false)
+	return hash
+}
+
+func (z *ZkMerkleStateTree) Commit(_ bool) (common.Hash, *trienode.NodeSet, error) {
+	err := z.ComputeAllNodeHash(func(node zk.TreeNode) error { return z.db.Put(node.Hash()[:], node.CanonicalValue()) })
+	if err != nil {
+		log.Error("Failed to commit zk merkle trie", "err", err)
+	}
+	// There is a bug where root node is saved twice.
+	// It is because of the bottom rawdb.WriteLegacyTrieNode, and we will remove it after checking if it can be removed.
+	rawdb.WriteLegacyTrieNode(z.db.diskdb, common.BytesToHash(z.RootNode().Hash().Bytes()), z.RootNode().CanonicalValue())
+	// Since NodeSet relies directly on mpt, we can't create a NodeSet.
+	// Of course, we might be able to force-fit it by implementing the node interface.
+	// However, NodeSet has been improved in geth, it could be improved to return a NodeSet when a commit is applied.
+	// So let's not do that for the time being.
+	// related geth commit : https://github.com/ethereum/go-ethereum/commit/bbcb5ea37b31c48a59f9aa5fad3fd22233c8a2ae
+	return common.BytesToHash(z.RootNode().Hash().Bytes()), nil, nil
+}
+
+func (z *ZkMerkleStateTree) NodeIterator(startKey []byte) (NodeIterator, error) {
```

Comment:
This function never returns an error, so do we need to return `(NodeIterator, error)` rather than just `NodeIterator`?

Reply:
I can't fix it because I tried to make the interface the same as the `NodeIterator` in `trie.Trie`

Reply:
Wouldn't it be possible to test snap sync by applying the contents of sync_test to the newly worked zk merkle tree?